Objective-C/CBLRemoteDatabase.mm

```diff
@@ -0,0 +1,219 @@
+    BOOL success = err == nil || err->code == 0;
+    NSError* error = nil;
+    if (!success && err) {
+        convertError(*err, &error);
+    }
+    
+    CBLDocument* updatedDoc = nil;
+    if (!ctx.isDeleted)
+        updatedDoc = [[CBLDocument alloc] initWithDocumentID: ctx.docID
+                                                  revisionID: slice2string(newRevID)
+                                                        body: ctx.docBody];
+    dispatch_async(ctx.remoteDB->_dispatchQueue, ^{
+        ctx.docUpdateCompletion(updatedDoc, error);
+        [ctx.remoteDB.contexts removeObject: ctx];
+    });
+}
+
+- (void) saveDocument: (CBLDocument*)document
+           asDeletion: (BOOL)deletion
+           completion: (void (^)(CBLDocument*, NSError*))completion {
+    
+    C4RevisionFlags revFlags = 0;
+    if (deletion)
+        revFlags = kRevDeleted;
+    FLSliceResult body;
+    if (!deletion && !document.isEmpty) {
+        // Encode properties to Fleece data:
+        // TODO: https://issues.couchbase.com/browse/CBL-2992
+        NSError* error = nil;
+        body = [document encodeWithRevFlags: &revFlags useSharedEncoder: NO error: &error];
+        if (!body.buf) {
+            completion(nil, error);
+            return;
+        }
+    } else {
+        FLEncoder enc = FLEncoder_New();
+        FLEncoder_BeginDict(enc, 0);
+        FLEncoder_EndDict(enc);
+        auto result = FLEncoder_Finish(enc, nullptr);
+        FLEncoder_Free(enc);
+        
+        body = result;
+    }
+    
+    // Update doc in the remote database
+    C4Error err = { };
+    CBLStringBytes docID(document.id);
+    CBLStringBytes revID(document.revisionID); // make sure, this is nullslice when no revisionID present
+    ConnectedClientUpdateDocumentContext* ctx;
+    if (!deletion)
+        ctx = [[ConnectedClientUpdateDocumentContext alloc] initWithRemoteDB: self
+                                                                       docID: document.id
+                                                                     docBody: body
+                                                                  completion: completion];
+    else
+        ctx = [[ConnectedClientUpdateDocumentContext alloc] initDeletionWithRemoteDB: self
+                                                                          completion: completion];
+    
+    [_contexts addObject: ctx];
+    BOOL success = c4client_putDoc(_client,
+                                   docID,
+                                   nullslice,
+                                   revID,
+                                   revFlags,
+                                   (FLSlice)body,
+                                   &updateDocumentCallback,
+                                   (__bridge void*)ctx,
+                                   &err);
+    
+    FLSliceResult_Release(body);
+    
+    if (!success || err.code != 0) {
```

Comment:
I understand that if !success, error code will not be zero. So if (!success) should be enough?

Reply:
Not sure, I found a crash with c4error nil, and it tries to convertError(). Assertion failure. Will add a `Assert()` to make sure, it crash in such case now and see.